### PR TITLE
Update isIssuedBy and isSignedBy to use util::StatusOr

### PR DESCRIPTION
**Pull request by Martin2112**
And corresponding change to isSelfSigned + update affected tests.

### CODE REVIEW

File: cpp/log/cert.cc [hunk 1]
```diff
@@ -402,14 +401,14 @@ Cert::Status Cert::IsSignedBy(const Cert& issuer) const {
          reason == ASN1_R_UNKNOWN_SIGNATURE_ALGORITHM)) {
       LOG(WARNING) << "Unsupported algorithm: " << PrintSignatureAlgorithm();
       ClearOpenSSLErrors();
-      return UNSUPPORTED_ALGORITHM;
+      return util::Status(Code::UNAVAILABLE, "Unsupported algorithm");
```
Comment (pphaneuf) [inline cpp/log/cert.cc]: Remember to make it `UNIMPLEMENTED` (and the other few spots).

File: cpp/log/cert_checker.cc [hunk 2]
```diff
@@ -306,26 +307,27 @@ Status CertChecker::GetTrustedCa(CertChain* chain) const {
        it != issuer_range.second; ++it) {
     const Cert* issuer_cand = it->second;
 
-    Cert::Status ok = subject->IsSignedBy(*issuer_cand);
-    if (ok == Cert::UNSUPPORTED_ALGORITHM) {
+    StatusOr<bool> signed_by_issuer = subject->IsSignedBy(*issuer_cand);
+    if (signed_by_issuer.status().error_code() == Code::UNAVAILABLE) {
```
Comment (pphaneuf) [inline cpp/log/cert_checker.cc]: Use `CanonicalCode()` instead of `error_code()`?

File: cpp/log/cert_test.cc [hunk 3]
```diff
@@ -351,7 +352,7 @@ TEST_F(CertTest, TestUnsupportedAlgorithm) {
 // see anything else. Make the test fail if this is not the case to notify the
 // user that their setup is insecure.
 #ifdef OPENSSL_NO_MD2
-  EXPECT_EQ(Cert::UNSUPPORTED_ALGORITHM, legacy.IsSignedBy(legacy));
+  EXPECT_EQ(Code::UNAVAILABLE, legacy.IsSignedBy(legacy).status().error_code());
```
Comment (pphaneuf) [inline cpp/log/cert_test.cc]: For `util::Status` tests, you can use `EXPECT_THAT`, with the `StatusIs` helper (from `cpp/util/status_test_util.h`, something like this:

```C++
EXPECT_THAT(legacy.IsSignedBy(legacy).status(), StatusIs(Code::UNAVAILABLE));
```

Yeah, they flip the expected/actual with `EXPECT_THAT`. Gotta keep it exciting! :smiley:

**Comment by pphaneuf** [comment]
All reviewed, mostly nits, really.

**Comment by Martin2112** [comment]
I think I've fixed everything. Also rebased and fixed up conflicts with other refactorings. It's starting to pay off a bit e.g look at CertChain::IsValidSignatureChain.

File: cpp/log/cert_checker.cc [hunk 4]
```diff
@@ -292,26 +293,27 @@ Status CertChecker::GetTrustedCa(CertChain* chain) const {
        it != issuer_range.second; ++it) {
     const Cert* issuer_cand = it->second;
 
-    Cert::Status ok = subject->IsSignedBy(*issuer_cand);
-    if (ok == Cert::UNSUPPORTED_ALGORITHM) {
+    StatusOr<bool> signed_by_issuer = subject->IsSignedBy(*issuer_cand);
+    if (signed_by_issuer.status().CanonicalCode() == Code::UNIMPLEMENTED) {
```
Comment (pphaneuf) [inline cpp/log/cert_checker.cc]: I wonder if we shouldn't just pass back the underlying status, or if in fact, this sort of translation is good (to control better what the end user see in the HTTP responses)?

**Comment by pphaneuf** [comment]
LGTM

File: cpp/log/cert_checker.cc (hunk 4 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: It's not an easy one to answer. In some cases the low level error can be
translated to something more meaningful at a higher level. OTOH you can end
up losing the actual cause of the problem by returning something generic.

**Pull request merged**